src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+﻿using System;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Logging;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using NuGet.Protocol.Core.v3;
+using System.Globalization;
+
+
+namespace NuGet.Protocol.Core.Types
+{
+    /// <summary>
+    /// Contains logics to push or delete packages in Http server or file system
+    /// </summary>
+    public class PackageUpdateResource : INuGetResource
+    {
+        private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+
+        private HttpSource _httpSource;
+        private string _source;
+
+        public PackageUpdateResource(string source,
+            HttpSource httpSource)
+        {
+            _source = source;
+            _httpSource = httpSource;
+        }
+
+        public async Task Push(string packagePath,
+            string source,
+            int timeoutInSecond,
+            Func<string, string> getApiKey,
+            ILogger log)
+        {
+            var timeout = TimeSpan.FromSeconds(Math.Abs(timeoutInSecond));
+            if (timeout.TotalSeconds == 0)
+            {
+                timeout = TimeSpan.FromMinutes(5); // Default to 5 minutes
+            }
+            var tokenSource = new CancellationTokenSource();
+            tokenSource.CancelAfter(timeout);
+
+            string apiKey = getApiKey(_source);
+
+            await PushPackage(packagePath, _source, apiKey, log, tokenSource.Token);
+
+            //likely values: https://www.nuget.org, https://www.nuget.org/api/v2/package, etc.
+            if (_source.StartsWith(NuGetConstants.DefaultGalleryServerUrl, StringComparison.OrdinalIgnoreCase))
```

Comment:
yes, check out the [code](https://github.com/NuGet/NuGet2/blob/2.11/src/CommandLine/Commands/PushCommand.cs#L50) here, I changed to the "startWith" to contain a few more varieties. I have no context for the original decision, but i can understand the symbols packages are for official release, while myget are still for on-going developments. we can still get rid of the restriction, but will it still makes senses to exclude it when push to file system?